noxfile.py

```diff
@@ -0,0 +1,84 @@
+import os
+from pathlib import Path
+
+import nox
+
+LINT_FILES = ("hacking/pr_labeler/label.py", "noxfile.py")
+PINNED = os.environ.get("PINNED", "true").lower() in {"1", "true"}
+nox.options.sessions = ("lint",)
+
+
+def install(session: nox.Session, *args, req: str, **kwargs):
+    arguments = ["-r", f"tests/{req}.in"]
+    if PINNED:
+        arguments.extend(("-c", f"tests/{req}.txt"))
+    session.install(
+        *arguments,
+        *args,
+        **kwargs,
+    )
+
+
+@nox.session
+def static(session: nox.Session):
+    """
+    Run static checkers
+    """
+    install(session, req="static")
+    session.run("ruff", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def formatters(session: nox.Session):
+    """
+    Reformat code
+    """
+    install(session, req="formatters")
+    session.run("isort", *session.posargs, *LINT_FILES)
+    session.run("black", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def formatters_check(session: nox.Session):
+    """
+    Check code formatting without making changes
+    """
+    install(session, req="formatters")
+    session.run("isort", "--check", *session.posargs, *LINT_FILES)
+    session.run("black", "--check", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def typing(session: nox.Session):
+    install(session, req="typing")
+    session.run("mypy", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def lint(session: nox.Session):
+    session.notify("static")
+    session.notify("formatters")
+
+
+requirements_files = list(
+    {path.name.replace(".in", "") for path in Path("tests").glob("*in")}
+    - {"constraints"}
+)
+
+
+@nox.session(name="pip-compile", python=["3.10"])
+@nox.parametrize(["req"], requirements_files, requirements_files)
+def pip_compile(session: nox.Session, req: str):
+    session.install("pip-tools")
+    # fmt: off
+    session.run(
+        "pip-compile",
+        "--resolver", "backtracking",
+        "--upgrade",
+        "--allow-unsafe",
+        "--quiet",
+        "--strip-extras",
+        "-o", f"tests/{req}.txt",
```

Comment:
Pro tip: this is how you can fight that black nonsense when it tries to treat key-val arguments separately.
```suggestion
        f"--output-file=tests/{req}.txt",
```

Reply:
I can change this to use the long option, as I generally think that's clearer, but the syntax you're proposing is even less readable IMO. I'm all for using linters and formatters, but I'd rather disable them when necessary than use less readable hacks to get around them.

Reply:
> I can change this to use the long option

Done

Reply:
> I'm all for using linters and formatters, but I'd rather disable them when necessary than use less readable hacks to get around them.

Well, I hate when formatters stand in the way, too. I mostly posted the suggestion as an educational comment rather than a request. So it's all up to you.